File: python/2326.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def spiralMatrix(self, m: int, n: int, head: Optional[ListNode]) -> List[List[int]]:
        ans = [[-1 for _ in range(n)] for _ in range(m)]

        t, b = 0, m - 1
        l, r = 0, n - 1

        while head:
            for j in range(l, r + 1):
                if not head: break
                ans[t][j] = head.val
                head = head.next
            
            for i in range(t + 1, b + 1):
                if not head: break
                ans[i][r] = head.val
                head = head.next
            
            for j in range(r - 1, l - 1, -1):
                if not head: break
                ans[b][j] = head.val
                head = head.next

            for i in range(b - 1, t, -1):
                if not head: break
                ans[i][l] = head.val
                head = head.next

            t += 1
            b -= 1
            l += 1
            r -= 1

        return ans

File: go/2326.go
package leetcode

type ListNode struct {
	Val  int
	Next *ListNode
}

func spiralMatrix(m int, n int, head *ListNode) [][]int {
	ans := make([][]int, m)
	for i := range ans {
		ans[i] = make([]int, n)
	}

	for i := range ans {
		for j := range ans[i] {
			ans[i][j] = -1
		}
	}

	t, b := 0, m-1
	l, r := 0, n-1

	for head != nil {
		for j := l; j < r+1; j++ {
			if head == nil {
				break
			}
			ans[t][j] = head.Val
			head = head.Next
		}
		for i := t + 1; i < b+1; i++ {
			if head == nil {
				break
			}
			ans[i][r] = head.Val
			head = head.Next
		}
		for j := r - 1; j > l-1; j-- {
			if head == nil {
				break
			}
			ans[b][j] = head.Val
			head = head.Next
		}
		for i := b - 1; i > t; i-- {
			if head == nil {
				break
			}
			ans[i][l] = head.Val
			head = head.Next
		}

		t++
		b--
		l++
		r--
	}

	return ans
}